File: gunkim/tony의문제집/Backtracking/14888_연산자끼워넣기.py
# 백트래킹의 느낌보다는 일반적인 dfs의 느낌이 강하게 난다.
n = int(input())
numbers = list(map(int, input().split()))
oper = list(map(int, input().split()))

min_value = float('inf')
max_value = -float('inf')


def calculate(index, sign, current_value):
    b = numbers[index+1]
    if sign == 0: # 덧셈
        return current_value + b
    elif sign == 1: # 뺄셈
        return current_value - b
    elif sign == 2: # 곱셉
        return current_value * b
    else: # 나눗셈
        quotient = abs(current_value) // abs(b)
        if current_value < 0:
            quotient *= -1
        if b < 0:
            quotient *= -1
        return quotient


def dfs(op, idx, result):
    if idx + 1 >= n: # 그나마 백트래킹? 그냥 종료 필수 조건이긴 함.
        global min_value, max_value
        min_value = min(min_value, result)
        max_value = max(max_value, result)
        return
    for i in range(4):
        if op[i] <= 0:
            continue
        op[i] -= 1
        dfs(op, idx+1, calculate(idx, i, result))
        op[i] += 1


for i in range(4):
    if oper[i] <= 0:
        continue
    oper[i] -= 1
    dfs(oper, 1, calculate(0, i, numbers[0]))
    oper[i] += 1

print(max_value)
print(min_value)

File: gunkim/today_problem/23_01_07/14888_연산자끼워넣기.java
import java.util.*;
import java.io.*;

public class Main {

    public int calculate(int n1, int n2, int opIdx) {
        if (opIdx == 0) {
            return n1 + n2;
        } else if (opIdx == 1) {
            return n1 - n2;
        } else if (opIdx == 2) {
            return n1 * n2;
        } else {
            return n1 / n2;
        }
    }

    public void dfs(int idx, int result) {
        if (idx >= n - 1) {
            minVal = Math.min(minVal, result);
            maxVal = Math.max(maxVal, result);
            return;
        }
        for (int i = 0; i < 4; i++) {
            if (operator[i] == 0) continue;
            operator[i] -= 1;
            dfs(idx + 1, calculate(result, nums[idx + 1], i));
            operator[i] += 1;
        }
    }

    public int n;
    public int[] nums;
    public int[] operator;
    public int minVal = Integer.MAX_VALUE;
    public int maxVal = Integer.MIN_VALUE;

    public void solution() throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        n = Integer.parseInt(br.readLine());
        nums = new int[n];
        StringTokenizer st = new StringTokenizer(br.readLine());
        for (int i = 0; i < n; i++) { //피연산자 초기화
            nums[i] = Integer.parseInt(st.nextToken());
        }
        operator = new int[4];
        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < 4; i++) { //연산자 수 초기화
            operator[i] = Integer.parseInt(st.nextToken());
        }

        for (int i = 0; i < 4; i++) { //dfs 탐색 시작
            if (operator[i] == 0) continue;
            operator[i] -= 1;
            dfs(1, calculate(nums[0], nums[1], i));
            operator[i] += 1;
        }

        System.out.println(maxVal);
        System.out.println(minVal);
    }

    public static void main(String[] args) throws Exception {
        new Main().solution();
    }

}